### ISSUE TITLE
ML should automagically handle when to conduct a top sort in a smarter way

### ISSUE THREAD

**Issue by baykaner** [fetchai/ledger#319]
for example, currently we can call predict first, and backprop second, and the graph constructed by predict will not include the loss function, such that the backprop process will fail.
